filebeat/input/file/state.go

```diff
@@ -31,6 +32,16 @@ func NewState(fileInfo os.FileInfo, path string) State {
 	}
 }
 
+// String returns a unique id for the state as a string
+func (s *State) String() string {
+	return s.FileStateOS.String()
+}
+
+// IsEqual compares the state to an other state supporing stringer based on the unique string
+func (s *State) IsEqual(c fmt.Stringer) bool {
+	return s.String() == c.String()
```

Comment:
Somehow relying on `String()` functions seems dangerous to me. Perhaps you could have the same logic, but based on a string `id` that is only computed once and stored in the state?